File: ios/Oriveo/Oriveo/Core/Providers/ProviderSetupCatalog.swift
import Foundation
import os

struct ProviderSetupDefaults: Sendable, Equatable {
    let displayName: String
    let shortName: String
    let apiKeyPlaceholder: String
    let defaultBaseURLText: String?
    let autoFillNote: String?
}

struct ProviderSetupCatalog: Sendable, Equatable {
    let directProviders: [ProviderKind]
    let aggregatorProviders: [ProviderKind]
    private let defaultsByKind: [ProviderKind: ProviderSetupDefaults]
    private let regionsByKind: [ProviderKind: [ProviderEndpointOption]]

    func displayName(for kind: ProviderKind) -> String {
        defaultsByKind[kind]?.displayName ?? kind.displayName
    }

    func shortName(for kind: ProviderKind) -> String {
        defaultsByKind[kind]?.shortName ?? kind.shortName
    }

    func apiKeyPlaceholder(for kind: ProviderKind) -> String {
        defaultsByKind[kind]?.apiKeyPlaceholder ?? kind.apiKeyPlaceholder
    }

    func defaultBaseURLText(for kind: ProviderKind) -> String? {
        defaultsByKind[kind]?.defaultBaseURLText ?? kind.defaultBaseURLText
    }

    func autoFillNote(for kind: ProviderKind) -> String? {
        defaultsByKind[kind]?.autoFillNote ?? kind.autoFillNote
    }

    func tagline(for kind: ProviderKind) -> String? {
        switch kind {
        case .openAI: return "Flagship GPT models"
        case .anthropic: return "Claude models"
        case .gemini: return "Text, image & video"
        case .deepseek: return "Reasoning & chat"
        case .grok: return "xAI models"
        case .miniMax: return "Multimodal models"
        case .zhipu: return "GLM models"
        case .qwen: return "Alibaba models"
        case .moonshot: return "Long context"
        case .mistral: return "Mistral AI models"
        case .siliconFlow: return "Multi-provider gateway"
        case .openRouter: return "200+ models"
        case .groq: return "Ultra-fast inference"
        case .together: return "Open-source models"
        case .fireworks: return "Fast & affordable"
        default: return nil
        }
    }

    func setupEndpointOptions(for kind: ProviderKind) -> [ProviderEndpointOption] {
        regionsByKind[kind] ?? kind.setupEndpointOptions
    }

    func defaultSetupEndpointID(for kind: ProviderKind) -> String? {
        setupEndpointOptions(for: kind).first?.id
    }

    func usesConfigurableBaseURL(_ kind: ProviderKind) -> Bool {
        !setupEndpointOptions(for: kind).isEmpty || kind == .relay
    }

    func resolvedSetupBaseURLText(for kind: ProviderKind, optionID: String?) -> String? {
        let normalizedOptionID = optionID?.trimmingCharacters(in: .whitespacesAndNewlines)
        if let normalizedOptionID,
           let matched = setupEndpointOptions(for: kind).first(where: { $0.id == normalizedOptionID }) {
            return matched.baseURLText
        }
        return setupEndpointOptions(for: kind).first?.baseURLText ?? defaultBaseURLText(for: kind)
    }

    func resolvedSetupEndpointOption(for kind: ProviderKind, baseURLText: String?) -> ProviderEndpointOption? {
        let options = setupEndpointOptions(for: kind)
        guard !options.isEmpty else { return nil }
        guard let normalized = Self.normalizedBaseURLText(baseURLText) else {
            return options.first
        }
        return options.first(where: {
            Self.normalizedBaseURLText($0.baseURLText) == normalized
        }) ?? options.first
    }

    private static let cachedCurrent =
        OSAllocatedUnfairLock<(generation: UInt64, catalog: ProviderSetupCatalog)?>(initialState: nil)

    static func current(metadata: MetadataClient = .shared) -> ProviderSetupCatalog {
        guard metadata === MetadataClient.shared else {
            return build(from: metadata)
        }
        let generation = MetadataClient.sharedSnapshotGeneration()
        if let cached = cachedCurrent.withLock({ $0 }), cached.generation == generation {
            return cached.catalog
        }
        let catalog = build(from: metadata)
        cachedCurrent.withLock { $0 = (generation, catalog) }
        return catalog
    }

    private static func build(from metadata: MetadataClient) -> ProviderSetupCatalog {
        fromProviderConfigs(
            metadata.syncHasPublicProviderConfigSource()
                ? metadata.syncListPublicProviderConfigs()
                : nil
        )
    }

    static func fromProviderConfigs(_ providerConfigs: [MetadataClient.PublicProviderConfig]?) -> ProviderSetupCatalog {
        guard let providerConfigs else { return fallback }

        let knownConfigs = providerConfigs
            .compactMap(KnownProviderConfig.init)
            .sorted { left, right in
                let leftOrder = left.sortOrder ?? Int.max
                let rightOrder = right.sortOrder ?? Int.max
                if leftOrder != rightOrder { return leftOrder < rightOrder }
                let leftIndex = fallbackProviderOrder.firstIndex(of: left.kind) ?? Int.max
                let rightIndex = fallbackProviderOrder.firstIndex(of: right.kind) ?? Int.max
                if leftIndex != rightIndex { return leftIndex < rightIndex }
                return left.kind.rawValue < right.kind.rawValue
            }

        var defaults = fallbackDefaults
        var regions = fallbackRegions
        var direct: [ProviderKind] = []
        var aggregators: [ProviderKind] = []

        for config in knownConfigs {
            defaults[config.kind] = ProviderSetupDefaults(
                displayName: config.selectionLabel ?? config.displayName,
                shortName: config.shortName ?? config.displayName,
                apiKeyPlaceholder: config.apiKeyPlaceholder ?? config.kind.apiKeyPlaceholder,
                defaultBaseURLText: config.defaultBaseURLText,
                autoFillNote: config.autoFillNote
            )
            if !config.regionOptions.isEmpty {
                regions[config.kind] = config.regionOptions
            }
            if config.category == "aggregator" {
                aggregators.append(config.kind)
            } else {
                direct.append(config.kind)
            }
        }

        return ProviderSetupCatalog(
            directProviders: direct,
            aggregatorProviders: aggregators,
            defaultsByKind: defaults,
            regionsByKind: regions
        )
    }

    static let fallback = ProviderSetupCatalog(
        directProviders: ProviderKind.directProviders,
        aggregatorProviders: ProviderKind.aggregatorProviders,
        defaultsByKind: fallbackDefaults,
        regionsByKind: fallbackRegions
    )

    private static let fallbackProviderOrder: [ProviderKind] = [
        .openAI,
        .anthropic,
        .gemini,
        .openRouter,
        .deepseek,
        .grok,
        .moonshot,
        .mistral,
        .siliconFlow,
        .groq,
        .together,
        .fireworks,
        .miniMax,
        .zhipu,
        .qwen,
    ]

    private static let fallbackDefaults: [ProviderKind: ProviderSetupDefaults] =
        Dictionary(uniqueKeysWithValues: fallbackProviderOrder.map { kind in
            (
                kind,
                ProviderSetupDefaults(
                    displayName: kind.displayName,
                    shortName: kind.shortName,
                    apiKeyPlaceholder: kind.apiKeyPlaceholder,
                    defaultBaseURLText: kind.defaultBaseURLText,
                    autoFillNote: kind.autoFillNote
                )
            )
        })

    private static let fallbackRegions: [ProviderKind: [ProviderEndpointOption]] =
        Dictionary(uniqueKeysWithValues: fallbackProviderOrder.map { ($0, $0.setupEndpointOptions) })

    private static func normalizedBaseURLText(_ baseURLText: String?) -> String? {
        guard var normalized = baseURLText?
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .lowercased(),
              !normalized.isEmpty else {
            return nil
        }
        if normalized.hasPrefix("https://") {
            normalized.removeFirst("https://".count)
        } else if normalized.hasPrefix("http://") {
            normalized.removeFirst("http://".count)
        }
        while normalized.hasSuffix("/") {
            normalized.removeLast()
        }
        return normalized
    }
}

private struct KnownProviderConfig {
    let kind: ProviderKind
    let displayName: String
    let shortName: String?
    let selectionLabel: String?
    let autoFillNote: String?
    let defaultBaseURLText: String?
    let apiKeyPlaceholder: String?
    let category: String?
    let regionOptions: [ProviderEndpointOption]
    let sortOrder: Int?

    nonisolated init?(_ config: MetadataClient.PublicProviderConfig) {
        guard let kind = Self.kind(from: config.kind) else { return nil }
        self.kind = kind
        displayName = config.displayName.trimmedNonEmpty ?? kind.displayName
        shortName = config.shortName?.trimmedNonEmpty
        selectionLabel = config.selectionLabel?.trimmedNonEmpty
        autoFillNote = config.autoFillNote?.trimmedNonEmpty
        defaultBaseURLText = Self.normalizedBaseURLText(config.defaultBaseURL)
        apiKeyPlaceholder = config.apiKeyPlaceholder?.trimmedNonEmpty
        category = (config.category == "direct" || config.category == "aggregator") ? config.category : nil
        regionOptions = (config.regionOptions ?? []).compactMap { option in
            guard let id = option.id.trimmedNonEmpty,
                  let label = option.label.trimmedNonEmpty,
                  let baseURLText = Self.normalizedBaseURLText(option.baseURL) else {
                return nil
            }
            return ProviderEndpointOption(id: id, label: label, baseURLText: baseURLText)
        }
        sortOrder = config.sortOrder
    }

    private static func kind(from raw: String) -> ProviderKind? {
        switch raw {
        case "openAI": return .openAI
        case "anthropic": return .anthropic
        case "gemini": return .gemini
        case "deepseek": return .deepseek
        case "grok": return .grok
        case "openRouter": return .openRouter
        case "groq": return .groq
        case "togetherAI": return .together
        case "fireworksAI": return .fireworks
        case "miniMax": return .miniMax
        case "zhipu": return .zhipu
        case "qwen": return .qwen
        case "moonshot": return .moonshot
        case "mistral": return .mistral
        case "siliconFlow": return .siliconFlow
        default: return nil
        }
    }

    private static func normalizedBaseURLText(_ raw: String) -> String? {
        var normalized = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !normalized.isEmpty else { return nil }
        if normalized.hasPrefix("https://") {
            normalized.removeFirst("https://".count)
        } else if normalized.hasPrefix("http://") {
            normalized.removeFirst("http://".count)
        }
        while normalized.hasSuffix("/") {
            normalized.removeLast()
        }
        return normalized
    }
}

private extension String {
    var trimmedNonEmpty: String? {
        let trimmed = trimmingCharacters(in: .whitespacesAndNewlines)
        return trimmed.isEmpty ? nil : trimmed
    }
}

File: android/app/src/main/java/ai/oriveo/community/feature/providers/setup/ProviderSetupCatalog.kt
package ai.oriveo.community.feature.providers.setup

import ai.oriveo.community.core.data.remote.MetadataClient
import ai.oriveo.community.core.model.ProviderKind
import ai.oriveo.community.core.model.RegionOption

data class ProviderSetupDefaults(
    val displayName: String,
    val shortName: String,
    val apiKeyPlaceholder: String,
    val defaultBaseUrl: String?,
    val autoFillNote: String?,
)

data class ProviderSetupCatalog(
    val directProviders: List<ProviderKind>,
    val aggregatorProviders: List<ProviderKind>,
    private val defaultsByKind: Map<ProviderKind, ProviderSetupDefaults>,
    private val regionsByKind: Map<ProviderKind, List<RegionOption>>,
) {
    fun displayName(kind: ProviderKind): String = defaultsByKind[kind]?.displayName ?: kind.displayName

    fun shortName(kind: ProviderKind): String = defaultsByKind[kind]?.shortName ?: kind.shortName

    fun apiKeyPlaceholder(kind: ProviderKind): String =
        defaultsByKind[kind]?.apiKeyPlaceholder ?: kind.apiKeyPlaceholder

    fun defaultBaseUrl(kind: ProviderKind): String? =
        defaultsByKind[kind]?.defaultBaseUrl ?: kind.defaultBaseUrl

    fun autoFillNote(kind: ProviderKind): String? =
        defaultsByKind[kind]?.autoFillNote ?: kind.autoFillNote

    fun regionOptions(kind: ProviderKind): List<RegionOption> =
        regionsByKind[kind] ?: kind.regionOptions

    fun resolveRegionOption(kind: ProviderKind, baseUrl: String?): RegionOption? {
        val options = regionOptions(kind)
        if (options.isEmpty()) return null
        val normalized = normalizeBaseUrl(baseUrl) ?: return options.firstOrNull()
        return options.firstOrNull { normalizeBaseUrl(it.baseURL) == normalized } ?: options.firstOrNull()
    }

    companion object {
        fun fromProviderConfigs(
            providerConfigs: List<MetadataClient.PublicProviderConfig>?,
        ): ProviderSetupCatalog {
            if (providerConfigs == null) return fallback()

            val knownConfigs = providerConfigs
                .mapNotNull { config -> config.toKnownProviderConfig() }
                .sortedWith(
                    compareBy<KnownProviderConfig> { it.sortOrder ?: Int.MAX_VALUE }
                        .thenBy { fallbackOrder.indexOf(it.kind).takeIf { index -> index >= 0 } ?: Int.MAX_VALUE }
                        .thenBy { it.kind.rawValue },
                )

            if (knownConfigs.isEmpty()) {
                return ProviderSetupCatalog(
                    directProviders = emptyList(),
                    aggregatorProviders = emptyList(),
                    defaultsByKind = fallbackDefaults(),
                    regionsByKind = fallbackRegions(),
                )
            }

            val defaults = fallbackDefaults().toMutableMap()
            val regions = fallbackRegions().toMutableMap()
            val direct = mutableListOf<ProviderKind>()
            val aggregators = mutableListOf<ProviderKind>()

            knownConfigs.forEach { config ->
                defaults[config.kind] = ProviderSetupDefaults(
                    displayName = config.selectionLabel ?: config.displayName,
                    shortName = config.shortName ?: config.displayName,
                    apiKeyPlaceholder = config.apiKeyPlaceholder ?: config.kind.apiKeyPlaceholder,
                    defaultBaseUrl = config.defaultBaseURL,
                    autoFillNote = config.autoFillNote,
                )
                
                if (config.regionOptions.isNotEmpty()) {
                    regions[config.kind] = config.regionOptions
                }
                when (config.category) {
                    "aggregator" -> aggregators.add(config.kind)
                    else -> direct.add(config.kind)
                }
            }

            return ProviderSetupCatalog(
                directProviders = direct,
                aggregatorProviders = aggregators,
                defaultsByKind = defaults,
                regionsByKind = regions,
            )
        }

        fun fallback(): ProviderSetupCatalog = ProviderSetupCatalog(
            directProviders = ProviderKind.directProviders,
            aggregatorProviders = ProviderKind.aggregators,
            defaultsByKind = fallbackDefaults(),
            regionsByKind = fallbackRegions(),
        )

        private val fallbackOrder = listOf(
            ProviderKind.OpenAI,
            ProviderKind.Anthropic,
            ProviderKind.Gemini,
            ProviderKind.OpenRouter,
            ProviderKind.DeepSeek,
            ProviderKind.Grok,
            ProviderKind.Moonshot,
            ProviderKind.Mistral,
            ProviderKind.SiliconFlow,
            ProviderKind.Groq,
            ProviderKind.Together,
            ProviderKind.Fireworks,
            ProviderKind.MiniMax,
            ProviderKind.Zhipu,
            ProviderKind.Qwen,
        )

        private fun fallbackDefaults(): Map<ProviderKind, ProviderSetupDefaults> =
            fallbackOrder.associateWith { kind ->
                ProviderSetupDefaults(
                    displayName = kind.displayName,
                    shortName = kind.shortName,
                    apiKeyPlaceholder = kind.apiKeyPlaceholder,
                    defaultBaseUrl = kind.defaultBaseUrl,
                    autoFillNote = kind.autoFillNote,
                )
            }

        private fun fallbackRegions(): Map<ProviderKind, List<RegionOption>> =
            fallbackOrder.associateWith { kind -> kind.regionOptions }
    }
}

object ProviderSetupCatalogResolver {
    fun current(): ProviderSetupCatalog = ProviderSetupCatalog.fromProviderConfigs(
        if (MetadataClient.hasPublicProviderConfigSource()) {
            MetadataClient.listPublicProviderConfigs()
        } else {
            null
        },
    )
}

private data class KnownProviderConfig(
    val kind: ProviderKind,
    val displayName: String,
    val shortName: String?,
    val selectionLabel: String?,
    val autoFillNote: String?,
    val defaultBaseURL: String?,
    val apiKeyPlaceholder: String?,
    val category: String?,
    val regionOptions: List<RegionOption>,
    val sortOrder: Int?,
)

private fun MetadataClient.PublicProviderConfig.toKnownProviderConfig(): KnownProviderConfig? {
    val kind = when (kind) {
        "openAI" -> ProviderKind.OpenAI
        "anthropic" -> ProviderKind.Anthropic
        "gemini" -> ProviderKind.Gemini
        "deepseek" -> ProviderKind.DeepSeek
        "grok" -> ProviderKind.Grok
        "openRouter" -> ProviderKind.OpenRouter
        "groq" -> ProviderKind.Groq
        "togetherAI" -> ProviderKind.Together
        "fireworksAI" -> ProviderKind.Fireworks
        "miniMax" -> ProviderKind.MiniMax
        "zhipu" -> ProviderKind.Zhipu
        "qwen" -> ProviderKind.Qwen
        "moonshot" -> ProviderKind.Moonshot
        "mistral" -> ProviderKind.Mistral
        "siliconFlow" -> ProviderKind.SiliconFlow
        else -> return null
    }
    val normalizedDefault = defaultBaseURL.trim().takeIf { it.isNotEmpty() }
    return KnownProviderConfig(
        kind = kind,
        displayName = displayName.trim().takeIf { it.isNotEmpty() } ?: kind.displayName,
        shortName = shortName?.trim()?.takeIf { it.isNotEmpty() },
        selectionLabel = selectionLabel?.trim()?.takeIf { it.isNotEmpty() },
        autoFillNote = autoFillNote?.trim()?.takeIf { it.isNotEmpty() },
        defaultBaseURL = normalizedDefault,
        apiKeyPlaceholder = apiKeyPlaceholder?.trim()?.takeIf { it.isNotEmpty() },
        category = category?.takeIf { it == "direct" || it == "aggregator" },
        regionOptions = regionOptions
            ?.mapNotNull { option ->
                val id = option.id.trim()
                val label = option.label.trim()
                val baseURL = option.baseURL.trim().trimEnd('/')
                if (id.isEmpty() || label.isEmpty() || baseURL.isEmpty()) {
                    null
                } else {
                    RegionOption(id, label, baseURL)
                }
            }
            ?: emptyList(),
        sortOrder = sortOrder,
    )
}

private fun normalizeBaseUrl(baseUrl: String?): String? {
    var normalized = baseUrl?.trim()?.lowercase().orEmpty()
    if (normalized.isEmpty()) return null
    normalized = normalized.removePrefix("https://").removePrefix("http://")
    while (normalized.endsWith("/")) {
        normalized = normalized.dropLast(1)
    }
    return normalized
}
